opentelemetry/proto/profiles/v1development/eventprofile.proto

```diff
@@ -0,0 +1,230 @@
+// for the fields time, duration, value, count and sum when considering that
+// they can be unset (0) or holding a value (>1). We need to be clear which
+// combinations have use cases and which should be considered invalid.
+message Event {
+  // The index of EventType associated with this event. The most frequent event
+  // type should use index 0 so it can be omitted on the wire.
+  uint32 event_type_index = 1;
+
+  // The index of the Stack associated with this event.
+  uint32 stack_index = 2;
+
+  // The index of the Link associated with this event if available.
+  optional uint32 link_index = 3;
+
+  // The indices of the attributes associated with this event.
+  repeated uint32 attribute_indices = 4;
+
+  // time represents the timestamp of the event. For events without a duration,
+  // this is the end time of the event. Time values for the same Clock must be
+  // delta encoded. This means that the time value of the first Event on the
+  // wire is the number of clock cycles since the start of the associated Clock.
+  // For subsequent Event messages for the same Clock, the time value is the
+  // number of clock cycles since the previous message on the wire.
+  int64 time = 5;
+
+  // The duration of the event. This is typically used for Off-CPU wait events.
+  // CPU samples are point in time events that have a value representing their
+  // magnitude, but a duration of 0. the time period during which the CPU
+  // time acrued is explicitly known. The unit of the duration is defined by the
+  // Clock associated with the EventType. When no Clock is associated with the
+  // the EventType, this should be 0 (omitted).
+  uint64 duration = 6;
+
+  // The value of the event in the unit specified in the EventType. If there is
+  // a Clock for the EventType, the value defaults to EventType.default_value
+  // when it is 0 (omitted). If there is no Clock for the EventType, the value
+  // defaults to Event.count * EventType.default_value. The latter case allows
+  // for the efficient aggregation of CPU samples with equal magnitudes.
+  uint64 value = 7;
+
+  // The number of events that were aggregated into this event. A count value of
+  // 0 should be used for individual events to reduce message size, but 1 is
+  // also allowed and considered to be equivalent. A count larger than 1
+  // indicates that this event is an aggregate of multiple events.
+  uint64 count = 8;
+
+  // The the sum the values of the aggregated events. If an event has a count
+  // and a EventType.default_value, the sum field should be 0, and clients
+  // should derive it by multiplying the count with the default value.
+  uint64 sum = 9;
+
+  // TODO: Add more statistics for aggregated events (e.g. min, max,
+  // stddev, etc) or even histogram sketches?
+}
+
+// Clock describes the clock format used by the EventType. Clocks are expected
+// to be monotonic and include the time when a system is suspended.
+message Clock {
```

Comment:
I like the idea of having the clock type to allow configurable and different resolution of timestamps in the same or different profiles, but one downside is - does it make simple and typical case more verbose? E.g. perhaps many collectors might end up just using Unix epoch microseconds with no attributes, but they will have to configure at least one clock.

Reply:
Yeah, it's a tradeoff. I added the clocks for a few different reasons. If I had to put them in order, it would probably be:

1. **Leap Seconds**: IMO observability systems should avoid unix epoch timestamps until leap seconds are [abolished in 2035](https://www.theverge.com/2022/11/21/23470759/leap-second-scrapped-2035-time-computer-chaos). At least I would imagine that they make it significantly less fun to debug incidents that are caused by leap seconds. But unfortunately this is a problem for all other OpenTelemetry signals already.
2. **Data volumes**: Changing the resolution could save 1-2 bytes per event under varint encoding.
3.  **TSC Clocks:** Those can run at frequencies above 1 Ghz, so nanosecond resolution is not sufficient to store their values. OTOH, I can think of very few use cases that require sub-nanosecond resolution, especially for the stuff we target with OpenTelemetry.

That being said, the first problem is solvable without configurable clocks. If we decide 2 and 3 are not important enough, we should probably just require timestamps that originate from a non-leap unix epoch with nanosecond precision?

Reply:
2 and 3 could be solved by having some way of specifying the resolution of the timestamp per event type? E.g. nanos, micros, or millis.

Reply:
We should probably look if all other signals follow some specific approach with this. Being extra special is probably not worth it.